src/platforms/stlinkv3/platform.c

```diff
@@ -0,0 +1,259 @@
+ * Portions (C) 2020-2021 Stoyan Shopov <stoyan.shopov@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file provides the platform specific functions for the ST-Link V3
+ * implementation.
+ */
+
+#include "general.h"
+#include "usb.h"
+#include "aux_serial.h"
+#include "gdb_if.h"
+
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/cm3/scb.h>
+#include <libopencm3/cm3/scs.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/usart.h>
+#include <libopencm3/usb/usbd.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/spi.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/syscfg.h>
+
+uint16_t srst_pin;
+static uint32_t hw_version;
+
+#define SCB_CCR_IC_Pos 17U                     /*!< SCB CCR: Instruction cache enable bit Position */
+#define SCB_CCR_IC_Msk (1UL << SCB_CCR_IC_Pos) /*!< SCB CCR: Instruction cache enable bit Mask */
+
+#define SCB_CCR_DC_Pos 16U                     /*!< SCB CCR: Cache enable bit Position */
+#define SCB_CCR_DC_Msk (1UL << SCB_CCR_DC_Pos) /*!< SCB CCR: DC Mask */
+
+#define SCB_CCSIDR_NUMSETS_Msk (0x7fffUL << SCB_CCSIDR_NUMSETS_Pos) /*!< SCB CCSIDR: NumSets Mask */
+#define SCB_CCSIDR_NUMSETS_Pos 13U                                  /*!< SCB CCSIDR: NumSets Position */
+
+#define SCB_CCSIDR_ASSOCIATIVITY_Pos 3U /*!< SCB CCSIDR: Associativity Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_Msk (0x3ffUL << SCB_CCSIDR_ASSOCIATIVITY_Pos) /*!< SCB CCSIDR: Associativity Mask */
+#define CCSIDR_WAYS(x)               (((x)&SCB_CCSIDR_ASSOCIATIVITY_Msk) >> SCB_CCSIDR_ASSOCIATIVITY_Pos)
+
+#define SCB_DCISW_SET_Pos 5U                             /*!< SCB DCISW: Set Position */
+#define SCB_DCISW_SET_Msk (0x1ffUL << SCB_DCISW_SET_Pos) /*!< SCB DCISW: Set Mask */
+
+#define SCB_DCISW_WAY_Pos 30U                        /*!< SCB DCISW: Way Position */
+#define SCB_DCISW_WAY_Msk (3UL << SCB_DCISW_WAY_Pos) /*!< SCB DCISW: Way Mask */
+
+#define CCSIDR_SETS(x) (((x)&SCB_CCSIDR_NUMSETS_Msk) >> SCB_CCSIDR_NUMSETS_Pos)
+
+static void __DSB(void)
+{
+	__asm__ volatile("dsb 0xf" ::: "memory");
+}
+
+static void __ISB(void)
+{
+	__asm__ volatile("isb 0xf" ::: "memory");
+}
+
+static void SCB_EnableICache(void)
+{
+	__DSB();
+	__ISB();
+	SCB_ICIALLU = 0UL; /* invalidate I-Cache */
+	__DSB();
+	__ISB();
+	SCB_CCR |= (uint32_t)SCB_CCR_IC_Msk; /* enable I-Cache */
```

Comment:
This cast should not be required, please remove it (this goes for other similar lines below). The macro should be all-caps (eg, `SCB_CCR_IC_MASK`)

Reply:
Did the cast here `(uint32_t)` wind up being required? we'd prefer to remove the casts if at all possible